### PR TITLE
feat: add support for async functions to preprocessor option

### LINKED ISSUE #264: Allow preprocessor to be async
- Operating System: any
- Node Version: any
- NPM Version: any
- webpack Version: any
- html-loader Version: 1

### Feature Proposal

Some preprocessors supportes async mode, why do not support it?

### Feature Use Case

Above

### PR DESCRIPTION
**Pull request by EslamHiko**
This PR contains a:

- [ ] **bugfix**
- [x] new **feature**
- [ ] **code refactor**
- [ ] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

fixes #264

### Breaking Changes
No

### CODE REVIEW

**Comment by evilebottnawi** [comment]
@EslamHiko Let's update documentation too

File: README.md [hunk 1]
```diff
@@ -50,12 +50,12 @@ module.exports = {
...
-| **[`preprocessor`](#preprocessor)** |    `{Function}`     |                 `undefined`                  | Allows pre-processing of content before handling |
-|     **[`minimize`](#minimize)**     | `{Boolean\|Object}` | `true` in production mode, otherwise `false` | Tell `html-loader` to minimize HTML              |
-|     **[`esModule`](#esmodule)**     |     `{Boolean}`     |                   `false`                    | Use ES modules syntax                            |
+|                Name                 |            Type             |                   Default                    |                   Description                    |
+| :---------------------------------: | :-------------------------: | :------------------------------------------: | :----------------------------------------------: |
+|   **[`attributes`](#attributes)**   |     `{Boolean\|Object}`     |                    `true`                    |       Enables/Disables attributes handling       |
+| **[`preprocessor`](#preprocessor)** | `{Function\|AsyncFunction}` |                 `undefined`                  | Allows pre-processing of content before handling |
```
Comment (evilebottnawi) [inline README.md]: No need `AsyncFunction` here

File: README.md [hunk 2]
```diff
@@ -387,6 +391,45 @@ module.exports = {
...
+
+            return result;
+          },
+        },
+      },
+    ],
+  },
+};
+```
+
```
Comment (evilebottnawi) [inline README.md]: No need `AsyncFunction` header, just add somehintg like that: `You can use async function, for example:`

File: README.md [hunk 3]
```diff
@@ -391,9 +391,8 @@ module.exports = {
 };
 ```
 
-#### `AsyncFunction`
-
-You can set the `preprocessor` option as an Asynchronous Function (`AsyncFunction`) instance.
+You can also set the `preprocessor` option as an Asynchronous Function (`async Function`) instance.
+For example :
```
Comment (evilebottnawi) [inline README.md]: Avoid space after `:`

File: README.md [hunk 4]
```diff
@@ -391,9 +391,8 @@ module.exports = {
 };
 ```
 
-#### `AsyncFunction`
-
-You can set the `preprocessor` option as an Asynchronous Function (`AsyncFunction`) instance.
+You can also set the `preprocessor` option as an Asynchronous Function (`async Function`) instance.
```
Comment (evilebottnawi) [inline README.md]: `as an asynchronous function instance.`

File: README.md [hunk 5]
```diff
@@ -50,12 +50,12 @@ module.exports = {
...
+| :---------------------------------: | :-----------------: | :------------------------------------------: | :----------------------------------------------: |
+|   **[`attributes`](#attributes)**   | `{Boolean\|Object}` |                    `true`                    |       Enables/Disables attributes handling       |
 | **[`preprocessor`](#preprocessor)** |    `{Function}`     |                 `undefined`                  | Allows pre-processing of content before handling |
-|     **[`minimize`](#minimize)**     | `{Boolean\|Object}` | `true` in production mode, otherwise `false` | Tell `html-loader` to minimize HTML              |
-|     **[`esModule`](#esmodule)**     |     `{Boolean}`     |                   `false`                    | Use ES modules syntax                            |
+|     **[`minimize`](#minimize)**     | `{Boolean\|Object}` | `true` in production mode, otherwise `false` |       Tell `html-loader` to minimize HTML        |
+|     **[`esModule`](#esmodule)**     |     `{Boolean}`     |                   `false`                    |              Use ES modules syntax               |
```
Comment (evilebottnawi) [inline README.md]: Why it is changed, if not new?

Comment (EslamHiko) [reply README.md]: Oh I added `:` in the last of --- under the description table which made the lint centralize the description 😅

**Pull request merged**